extensions/stackdriver/stackdriver.cc

```diff
@@ -151,12 +151,45 @@ std::string getMonitoringEndpoint() {
   return monitoring_endpoint;
 }
 
+// Get GCP project number.
+std::string getProjectNumber() {
+  std::string project_number;
+  if (!getValue({"node", "metadata", "PLATFORM_METADATA", kGCPProjectNumberKey},
+                &project_number)) {
+    return "";
+  }
+  return project_number;
+}
+
 void clearTcpMetrics(::Wasm::Common::RequestInfo& request_info) {
   request_info.tcp_connections_opened = 0;
   request_info.tcp_sent_bytes = 0;
   request_info.tcp_received_bytes = 0;
 }
 
+// Get local node metadata. If mesh id is not filled or does not exist,
+// fall back to default format `proj-<project-number>`.
+void getLocalNodeMetadata(google::protobuf::Struct* node_metadata) {
+  if (!getMessageValue({"node", "metadata"}, node_metadata)) {
+    return;
+  }
+  const auto mesh_id_it = node_metadata->fields().find("MESH_ID");
+  if (mesh_id_it != node_metadata->fields().end() &&
+      !mesh_id_it->second.string_value().empty()) {
+    return;
+  }
+
+  // Insert or update mesh id to default mesh id if empty.
+  auto project_number = getProjectNumber();
+  auto* mesh_id_field =
+      (*node_metadata->mutable_fields())["MESH_ID"].mutable_string_value();
+  if (project_number.empty()) {
+    *mesh_id_field = "unknown";
```

Comment:
What is the difference between meshid = unknown and unset?

Reply:
Actually not much, I think both edge and metric handle this case well (if empty, set it to unknown). For log, it currently leaves unknown fields empty, so this is going to be make a difference there.